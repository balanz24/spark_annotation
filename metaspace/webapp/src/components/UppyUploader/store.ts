// based on https://uppy.io/docs/stores/#Implementing-Stores

import { ref } from '@vue/composition-api'
import { State } from '@uppy/core'

type Listener = (prevState: State, nextState: State, patch: object) => void

export default () => {
  const state = ref<State>({})
  const listeners = new Set<Listener>()

  return {
    getState: () => state.value,
    setState: (patch: object) => {
      const prevState = state.value
      const nextState = { ...prevState, ...patch }

      state.value = nextState

      listeners.forEach(listener => {
        listener(prevState, nextState, patch)
      })
    },
    subscribe: (listener: Listener) => {
      listeners.add(listener)
      return () => listeners.delete(listener)
    },
  }
}
