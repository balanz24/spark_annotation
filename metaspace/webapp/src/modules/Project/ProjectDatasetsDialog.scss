.project-datasets-dialog {
  .el-dialog {
    @apply max-w-lg;

    .el-form-item {
      margin-bottom: 10px;
    }
    .el-form-item__label {
      line-height: 1.2em;
    }
  }

  .el-dialog__header {
    padding-bottom: 0;
  }

  .el-dialog__body {
    padding: 20px;
  }

  .dataset-checkbox-list {
    margin: 12px 0;
    max-height: 50vh;
    overflow: auto;
    box-shadow: inset 0 -10px 7px -10px lightgrey;
  }
  .select-buttons {
    margin: 12px 0;
  }
  .el-checkbox__label {
    @apply inline-flex flex-grow justify-between items-baseline overflow-hidden;
  }


  .button-bar {
    display: flex;
    flex-wrap: wrap;
    justify-content: flex-end;
    margin-top: 20px;
  }

  .select-link{
    @apply cursor-pointer underline;
    color: hsl(208, 93%, 44%);
  }

  .el-checkbox:last-of-type {
    @apply mx-2;
  }
}

