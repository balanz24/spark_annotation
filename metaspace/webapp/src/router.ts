/* eslint-disable vue/max-len */
import Vue, { AsyncComponent } from 'vue'
import VueRouter, { RawLocation } from 'vue-router'
import AboutPage from './modules/App/AboutPage'
import DatasetsPage from './modules/Datasets/DatasetsPage.vue'
import { DialogPage, ResetPasswordPage } from './modules/Account'
import { redirectAfterSignIn } from './modules/Account/signInReturnUrl'
import { updateDBParam } from './modules/Filters/url'
import NotFoundPage from './modules/App/NotFoundPage.vue'

Vue.use(VueRouter)

const asyncPagesFreelyTyped = {
  AnnotationsPage: () => import(/* webpackPrefetch: true, webpackChunkName: "AnnotationsPage" */ './modules/Annotations/AnnotationsPage.vue'),
  DatasetSummary: () => import(/* webpackPrefetch: true, webpackChunkName: "DatasetSummary" */ './modules/Datasets/summary/DatasetSummary.vue'),
  MetadataEditPage: () => import(/* webpackPrefetch: true, webpackChunkName: "MetadataEditPage" */ './modules/MetadataEditor/MetadataEditPage.vue'),
  ImageAlignmentPage: () => import(/* webpackPrefetch: true, webpackChunkName: "ImageAlignmentPage" */ './modules/ImageAlignment/ImageAlignmentPage.vue'),
  UploadPage: () => import(/* webpackPrefetch: true, webpackChunkName: "UploadPage" */ './modules/MetadataEditor/UploadPage.vue'),
  DatasetOverviewPage: () => import(/* webpackPrefetch: true, webpackChunkName: "DatasetOverviewPage" */ './modules/Datasets/overview/DatasetOverviewPage'),
  DatasetComparisonPage: () => import(/* webpackPrefetch: true, webpackChunkName: "DatasetOverviewPage" */ './modules/Datasets/comparison/DatasetComparisonPage'),
  DatasetBrowserPage: () => import(/* webpackPrefetch: true, webpackChunkName: "DatasetOverviewPage" */ './modules/Datasets/imzml/DatasetBrowserPage'),
  DatasetEnrichmentPage: () => import(/* webpackPrefetch: true, webpackChunkName: "DatasetOverviewPage" */ './modules/Datasets/enrichment/DatasetEnrichmentPage'),
  SpottingProjectPage: () => import(/* webpackPrefetch: true, webpackChunkName: "SpottingProjectPage" */ './modules/SpottingProject/DashboardPage'),

  // These pages are relatively small as they don't have any big 3rd party dependencies, so pack them together
  DatasetTable: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Datasets/list/DatasetTable.vue'),
  HelpPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/App/HelpPage.vue'),
  EditUserPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/UserProfile/EditUserPage.vue'),
  CreateGroupPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Group/CreateGroupPage.vue'),
  ProjectsListPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Project/ProjectsListPage.vue'),
  SystemHealthPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Admin/SystemHealthPage.vue'),
  GroupsListPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Group/GroupsListPage'),
  PrivacyPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/App/PrivacyPage.vue'),
  TermsPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/App/TermsPage.vue'),
  PublicationsPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/App/PublicationsPage'),

  // These pages use sanitizeHtml, which is big
  ViewGroupPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle2" */ './modules/Group/ViewGroupPage.vue'),
  ViewProjectPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle2" */ './modules/Project/ViewProjectPage.vue'),

  // Separate bundle for design docs
  DesignTOC: () => import(/* webpackChunkName: "DesignBundle" */ './design/TOCPage.vue'),
  DesignStyleGuide: () => import(/* webpackChunkName: "DesignBundle" */ './design/StyleGuidePage.vue'),
  DesignIcons: () => import(/* webpackChunkName: "DesignBundle" */ './design/IconsPage.vue'),
  DesignComponents: () => import(/* webpackChunkName: "DesignBundle" */ './design/ComponentsPage.vue'),
  DesignForms: () => import(/* webpackChunkName: "DesignBundle" */ './design/FormsPage.vue'),
}
const asyncPages = asyncPagesFreelyTyped as Record<keyof typeof asyncPagesFreelyTyped, AsyncComponent>

const convertLegacyUrls = () => {
  const { pathname, hash, search } = window.location
  if (pathname === '/' && hash && hash.startsWith('#/')) {
    history.replaceState(undefined, undefined as any, hash.slice(1))
  }
  if (pathname === '/annotations') {
    const updatedQueryString = updateDBParam(search.slice(1))
    if (updatedQueryString !== null) {
      history.replaceState(undefined, undefined as any, `${pathname}?${updatedQueryString}`)
    }
  }
}
convertLegacyUrls()

const router = new VueRouter({
  mode: 'history',
  routes: [
    { path: '/(about)?', component: AboutPage, meta: { footer: true, headerClass: 'bg-primary' } },
    { path: '/annotations', name: 'annotations', component: asyncPages.AnnotationsPage },
    {
      path: '/datasets',
      component: DatasetsPage,
      children: [
        { path: '', component: asyncPages.DatasetTable },
        { path: 'summary', component: asyncPages.DatasetSummary },
      ],
    },
    {
      path: '/datasets/:dataset_id/comparison',
      name: 'datasets-comparison',
      component: asyncPages.DatasetComparisonPage,
    },
    { path: '/datasets/edit/:dataset_id', name: 'edit-metadata', component: asyncPages.MetadataEditPage },
    { path: '/datasets/:dataset_id/add-optical-image', name: 'add-optical-image', component: asyncPages.ImageAlignmentPage },
    { path: '/dataset/:dataset_id', name: 'dataset-overview', component: asyncPages.DatasetOverviewPage },
    { path: '/dataset/:dataset_id/annotations', name: 'dataset-annotations', component: asyncPages.AnnotationsPage },
    { path: '/dataset/:dataset_id/browser', name: 'dataset-browser', component: asyncPages.DatasetBrowserPage },
    { path: '/dataset/:dataset_id/enrichment', name: 'dataset-enrichment', component: asyncPages.DatasetEnrichmentPage },
    { path: '/upload', component: asyncPages.UploadPage },
    { path: '/help', component: asyncPages.HelpPage, meta: { footer: true } },
    { path: '/user/me', component: asyncPages.EditUserPage },

    { path: '/admin/health', component: asyncPages.SystemHealthPage },

    { path: '/account/sign-in', component: DialogPage, props: { dialog: 'signIn' } },
    { path: '/account/sign-in-success', redirect: redirectAfterSignIn },
    { path: '/account/create-account', component: DialogPage, props: { dialog: 'createAccount' } },
    { path: '/account/forgot-password', component: DialogPage, props: { dialog: 'forgotPassword' } },
    { path: '/account/reset-password', component: ResetPasswordPage },

    { path: '/groups', component: asyncPages.GroupsListPage },
    { path: '/group/create', component: asyncPages.CreateGroupPage },
    { path: '/group/:groupIdOrSlug', name: 'group', component: asyncPages.ViewGroupPage },

    { path: '/project/:projectIdOrSlug', name: 'project', component: asyncPages.ViewProjectPage },
    { // Legacy URL sent in "request access" emails up until Feb 2019
      path: '/project/:projectIdOrSlug/manage',
      redirect: { path: '/project/:projectIdOrSlug', query: { tab: 'members' } } as RawLocation,
    },
    { path: '/projects', component: asyncPages.ProjectsListPage },

    { path: '/terms', component: asyncPages.TermsPage, meta: { footer: true } },
    { path: '/privacy', component: asyncPages.PrivacyPage, meta: { footer: true } },
    { path: '/publications', component: asyncPages.PublicationsPage, meta: { footer: true } },

    { path: '/design', component: asyncPages.DesignTOC, meta: { footer: true, flex: true } },
    { path: '/design/styleguide', component: asyncPages.DesignStyleGuide, meta: { footer: true, flex: true } },
    { path: '/design/icons', component: asyncPages.DesignIcons, meta: { footer: true, flex: true } },
    { path: '/design/components', component: asyncPages.DesignComponents, meta: { footer: true, flex: true } },
    { path: '/design/forms', component: asyncPages.DesignForms, meta: { footer: true, flex: true } },

    { path: '/detectability', name: 'detectability', component: asyncPages.SpottingProjectPage },

    { path: '*', component: NotFoundPage, meta: { footer: true, flex: true } },
  ],
  scrollBehavior(to, from, savedPosition) {
    if (to.hash) {
      return { selector: to.hash, offset: { x: 0, y: 64 } } // offset header height
    }
    if (savedPosition) {
      return savedPosition
    }
    if (to.path !== from.path) {
      return { x: 0, y: 0 }
    }
  },
})

const pageLoadedAt = Date.now()
router.beforeEach((to, from, next) => {
  // If user has had the window open for a long time, it's likely they're using old code. This can cause issues
  // such as incorrect API usage, or causing error reports for bugs that have already been fixed.
  // Force the browser to reload the page, preferring to do so during page transitions to minimize impact.
  const daysSincePageLoad = (Date.now() - pageLoadedAt) / 1000 / 60 / 60 / 24
  if (to.path !== from.path && daysSincePageLoad > 1.75 || daysSincePageLoad > 7) {
    window.location.assign(to.fullPath)
    next(false)
  } else {
    next()
  }
})

const { href } = router.resolve({ name: 'project', params: { projectIdOrSlug: 'REMOVE' } }, undefined, true)
export const PROJECT_URL_PREFIX = location.origin + href.replace('REMOVE', '')

export default router
