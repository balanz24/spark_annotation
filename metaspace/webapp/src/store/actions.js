import apolloClient from '../api/graphqlClient';
import {fetchOptionListsQuery} from '../api/metadata';
import {decodeParams} from '../modules/Filters';
import {computed} from "@vue/composition-api";
import store from "../store/index";
export default {

  async initFilterLists(context) {
    if(context.state.filterListsLoading || context.state.filterLists != null)
      return;

    context.commit('setFilterListsLoading');

    const response = await apolloClient.query({
      query: fetchOptionListsQuery,
      fetchPolicy: 'cache-first',
    });

    // set annotationIds default according to values passed from snapshot to the store state
    context.commit('setFilterLists', {...response.data,
      annotationIds: computed(() => store.state.snapshotAnnotationIds)} );

    // Refresh the current filter so that computed defaults that depend on `filterLists` are applied
    const filter = decodeParams(context.state.route, context.state.filterLists)
    context.commit('updateFilter', filter)
  },

};
