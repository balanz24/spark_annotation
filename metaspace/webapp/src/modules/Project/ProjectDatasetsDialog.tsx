import { defineComponent, computed, reactive } from '@vue/composition-api'
import { importDatasetsIntoProjectMutation, ProjectsListProject } from '../../api/project'
import { Dialog, Checkbox, Button } from '../../lib/element-ui'
import { uniqBy, isEmpty } from 'lodash'
import './ProjectDatasetsDialog.scss'
import { DatasetListItem, datasetListItemsQuery } from '../../api/dataset'
import { useQuery } from '@vue/apollo-composable'
import ElapsedTime from '../../components/ElapsedTime'
import reportError from '../../lib/reportError'
import Vue from 'vue'

interface CheckOptions {
  [key: string]: boolean
}

interface ProjectDatasetsDialogState {
  selectedDatasets: CheckOptions
  isSubmitting: boolean
  hasChanged: boolean
}

interface ProjectDatasetsDialogProps {
  project: ProjectsListProject
  currentUserId: string
  visible: boolean
  isManager: boolean
  refreshData: () => Promise<any>
}

export const ProjectDatasetsDialog = defineComponent<ProjectDatasetsDialogProps>({
  name: 'ProjectDatasetsDialog',
  props: {
    project: { type: Object, default: undefined },
    currentUserId: { type: String, default: undefined },
    visible: { type: Boolean, default: true },
    isManager: { type: Boolean, default: false },
    refreshData: { type: Function, required: true },
  },
  setup(props, ctx) {
    const { emit, root } = ctx
    const { $apollo } = root
    const state = reactive<ProjectDatasetsDialogState>({
      selectedDatasets: {},
      isSubmitting: false,
      hasChanged: false,
    })

    const getProjectDatasets = (projectDatasets: any) => {
      const defaultDatasets = {} as CheckOptions

      if (projectDatasets) {
        projectDatasets.forEach((ds: any) => {
          defaultDatasets[ds.id] = true
        })
      }
      return defaultDatasets
    }

    const getAvailableDatasets = (datasets: any, projectDatasets: any) => {
      if (!props.isManager) {
        return datasets
      }

      if (datasets && projectDatasets) {
        // show not owned datasets if admin, so it can be removed
        return uniqBy(datasets.concat(projectDatasets), 'id')
      }

      return []
    }

    const {
      result: datasetsResult,
      loading,
    } = useQuery<{allDatasets: DatasetListItem[]}>(datasetListItemsQuery,
      () => ({ dFilter: { submitter: props.currentUserId } }))
    const datasets = computed(() => datasetsResult.value != null
      ? datasetsResult.value.allDatasets as DatasetListItem[] : null)
    const {
      result: projectDatasetsResult,
      loading: loadingProjectDatasets,
      refetch: projectDatasetsRefetch,
    } = useQuery<{allDatasets: DatasetListItem[]}>(datasetListItemsQuery,
      () => ({ dFilter: { project: props.project?.id } }))
    const projectDatasets = computed(() => {
      if (isEmpty(state.selectedDatasets)) {
        state.selectedDatasets = getProjectDatasets(projectDatasetsResult.value?.allDatasets)
      }

      return projectDatasetsResult.value != null
        ? projectDatasetsResult.value.allDatasets as DatasetListItem[] : null
    })

    const handleClose = () => {
      emit('close')
    }

    const handleDatasetCheck = (value: boolean, key: string) => {
      Vue.set(state.selectedDatasets, key, value)
      if (!state.hasChanged) {
        state.hasChanged = true
      }
    }

    const handleSelectNone = () => {
      const availableDatasets = getAvailableDatasets(datasets?.value,
        projectDatasets?.value).map((ds: any) => ds.id)
      availableDatasets.forEach((key: string) => {
        Vue.set(state.selectedDatasets, key, false)
      })
      if (!state.hasChanged) {
        state.hasChanged = true
      }
    }

    const handleSelectAll = () => {
      const availableDatasets = getAvailableDatasets(datasets?.value,
        projectDatasets?.value).map((ds: any) => ds.id)
      availableDatasets.forEach((key: string) => {
        Vue.set(state.selectedDatasets, key, true)
      })
      if (!state.hasChanged) {
        state.hasChanged = true
      }
    }

    const handleProjectDatasetsUpdate = async() => {
      const previousDatasets = getProjectDatasets(projectDatasets?.value)
      const removedDatasetIds: string[] = []
      const addedDatasetIds: string[] = []
      state.isSubmitting = true
      Object.keys(state.selectedDatasets).forEach((key: string) => {
        if (previousDatasets[key] && !state.selectedDatasets[key]) { // remove ds
          removedDatasetIds.push(key)
        } else if (!previousDatasets[key] && state.selectedDatasets[key]) { // new ds * do not add already
          addedDatasetIds.push(key)
        }
      })

      try {
        if (addedDatasetIds.length > 0 || removedDatasetIds.length > 0) {
          await $apollo.mutate({
            mutation: importDatasetsIntoProjectMutation,
            variables: { projectId: props.project?.id, datasetIds: addedDatasetIds, removedDatasetIds },
          })
        }
        await props.refreshData()
        await projectDatasetsRefetch()
        emit('update')
      } catch (err) {
        reportError(err)
      } finally {
        state.isSubmitting = false
      }
    }

    return () => {
      const {
        visible, project,
      } = props
      const { name } = project
      const availableDatasets = getAvailableDatasets(datasets?.value, projectDatasets?.value)

      return (
        <Dialog
          class='project-datasets-dialog'
          visible={visible}
          append-to-body
          title={`Manage ${name}\`s datasets`}
          lockScroll={false}
          onClose={handleClose}>
          <div class="mt-6">
            <h4 class="m-0">
              Would you like to include/remove previously submitted datasets?
            </h4>
            {
              availableDatasets != null && availableDatasets.length > 0
                && <div class="dataset-checkbox-list leading-6">
                  <div class="mb-2">
                    <span
                      class="select-link"
                      onClick={handleSelectNone}>Select none</span>
                    <span> | </span>
                    <span
                      class="select-link"
                      onClick={handleSelectAll}>Select all</span>
                  </div>
                  {
                    availableDatasets.map((dataset: any) => <Checkbox
                      class="flex h-6 items-center m-0 mx-2"
                      key={dataset.id}
                      value={state.selectedDatasets[dataset.id]}
                      disabled={state.isSubmitting}
                      onChange={(value: boolean) => handleDatasetCheck(value, dataset.id)}
                      label={dataset.name}>
                      <span class="truncate">
                        { dataset?.name }
                      </span>
                      <span class="text-gray-700 text-xs tracking-wide pl-1">
                        <ElapsedTime date={dataset?.uploadDT} />
                      </span>
                    </Checkbox>)
                  }
                </div>
            }
            {
              availableDatasets !== null && availableDatasets.length === 0
                && <div class="flex items-center justify-center leading-6">
                  <div>No datasets available</div>
                </div>
            }

            <div class="button-bar">
              <Button onClick={handleClose}>
                Cancel
              </Button>
              <Button
                class='w-32'
                loading={state.isSubmitting}
                disabled={!state.hasChanged}
                type="primary"
                onClick={handleProjectDatasetsUpdate}>
                Update
              </Button>
            </div>
          </div>
        </Dialog>
      )
    }
  },
})
