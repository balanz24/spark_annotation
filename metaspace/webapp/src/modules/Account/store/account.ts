import { Module } from 'vuex'
import { DialogType } from '../dialogs'
import router from '../../../router'
import { RawLocation } from 'vue-router/types/router'

export interface AccountState {
  dialog: DialogType | null;
  dialogCloseRedirect: RawLocation | null;
  loginSuccessRedirect: RawLocation | null;
}

interface ShowDialogOptions {
  dialog: DialogType;
  dialogCloseRedirect?: RawLocation;
  loginSuccessRedirect?: RawLocation;
}

interface HideDialogOptions {
  dialog: DialogType;
  isLoginSuccess?: boolean;
}

const account: Module<AccountState, any> = {
  namespaced: true,
  state: {
    dialog: null,
    dialogCloseRedirect: null,
    loginSuccessRedirect: null,
  },
  mutations: {
    showDialog(state, payload: DialogType | ShowDialogOptions) {
      const options: ShowDialogOptions = typeof payload === 'string' ? { dialog: payload } : payload

      state.dialog = options.dialog
      if (options.dialogCloseRedirect != null) {
        state.dialogCloseRedirect = options.dialogCloseRedirect
      }
      if (options.loginSuccessRedirect != null) {
        state.loginSuccessRedirect = options.loginSuccessRedirect
      }
    },
    hideDialog(state, payload: DialogType | HideDialogOptions) {
      const options: HideDialogOptions = typeof payload === 'string' ? { dialog: payload } : payload

      if (state.dialog === options.dialog) {
        state.dialog = null
        if (options.isLoginSuccess && state.loginSuccessRedirect != null) {
          router.push(state.loginSuccessRedirect)
        } else if (state.dialogCloseRedirect != null) {
          router.push(state.dialogCloseRedirect)
        }

        state.dialogCloseRedirect = null
        state.loginSuccessRedirect = null
      }
    },
  },
}

export default account
