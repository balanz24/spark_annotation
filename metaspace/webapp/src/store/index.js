import Vue from 'vue';
import Vuex from 'vuex';
Vue.use(Vuex);

import getters from './getters.js';
import mutations from './mutations.js';
import actions from './actions.js';
import {accountModule} from '../modules/Account';

const store = new Vuex.Store({
  state: {
    // names of currently shown filters
    orderedActiveFilters: [],

    filterLists: null,
    filterListsLoading: false,

    // currently selected annotation
    annotation: undefined,

    // currently selected normalization matrix
    normalization: undefined,

    // roi settings
    roiInfo: {visible: false},

    // is annotation table loading?
    tableIsLoading: true,

    lastUsedFilters: {},

    currentTour: null,

    currentUser: {},

    // ion image global viewer settings
    channels: [],
    mode: 'SINGLE'
  },

  getters,
  mutations,
  actions,
  modules: {
    account: accountModule
  }
})

export default store;
