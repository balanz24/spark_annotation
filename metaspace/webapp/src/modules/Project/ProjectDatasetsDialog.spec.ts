import { mount } from '@vue/test-utils'
import router from '../../router'
import store from '../../store/index'
import Vue from 'vue'
import Vuex from 'vuex'
import { sync } from 'vuex-router-sync'
import { ProjectDatasetsDialog } from './ProjectDatasetsDialog'
import { initMockGraphqlClient, apolloProvider } from '../../../tests/utils/mockGraphqlClient'

Vue.use(Vuex)
sync(store, router)

describe('ProjectDatasetsDialog', () => {
  const propsData = {
    project: {
      id: '05c6519a-8049-11eb-927e-6bf28a9b25ae',
      name: 'Test',
      currentUserRole: 'MANAGER',
    },
    currentUserId: '039801c8-919e-11eb-908e-3b2b8e672707',
    visible: true,
    isManager: true,
    refreshData: () => {},
  }

  const testHarness = Vue.extend({
    components: {
      ProjectDatasetsDialog,
    },
    render(h) {
      return h(ProjectDatasetsDialog, { props: this.$attrs })
    },
  })

  const graphqlWithData = () => {
    initMockGraphqlClient({
      Query: () => ({
        allDatasets: () => {
          return [{
            id: '2021-03-31_11h02m28s',
            name: 'New 3 (1)',
            uploadDT: '2021-03-31T14:02:28.722Z',
          }, {
            id: '2021-03-30_18h25m18s',
            name: 'Untreated_3_434_super_lite_19_31 (1)',
            uploadDT: '2021-03-30T21:25:18.473Z',
          }]
        },
      }),
    })
  }
  const graphqlWithExtraUserData = () => {
    initMockGraphqlClient({
      Query: () => ({
        allDatasets: (src: any, { filter: { project } } : any) => {
          if (project) {
            return []
          }

          return [{
            id: '2021-03-31_11h02m28s',
            name: 'New 3 (1)',
            uploadDT: '2021-03-31T14:02:28.722Z',
          }, {
            id: '2021-03-30_18h25m18s',
            name: 'Untreated_3_434_super_lite_19_31 (1)',
            uploadDT: '2021-03-30T21:25:18.473Z',
          }]
        },
      }),
    })
  }

  const graphqlWithNoData = () => {
    initMockGraphqlClient({
      Query: () => ({
        allDatasets: () => {
          return []
        },
      }),
    })
  }

  it('it should match snapshot', async() => {
    graphqlWithData()
    const wrapper = mount(testHarness, { store, router, apolloProvider, propsData })
    await Vue.nextTick()

    expect(wrapper).toMatchSnapshot()
  })

  it('it should match no dataset snapshot', async() => {
    graphqlWithNoData()
    const wrapper = mount(testHarness, { store, router, apolloProvider, propsData })
    await Vue.nextTick()

    expect(wrapper).toMatchSnapshot()
  })

  it('it should have disabled update button when no data available', async() => {
    graphqlWithNoData()
    const wrapper = mount(testHarness, { store, router, apolloProvider, propsData })
    await Vue.nextTick()

    expect(wrapper.find('.el-button--primary').props('disabled')).toBe(true)
  })

  it('it should have disabled update button when no dataset selection has changed', async() => {
    graphqlWithData()
    const wrapper = mount(testHarness, { store, router, apolloProvider, propsData })
    await Vue.nextTick()

    expect(wrapper.find('.el-button--primary').props('disabled')).toBe(true)
  })

  it('it should uncheck all datasets when clicking select none', async() => {
    graphqlWithData()
    const wrapper = mount(testHarness, { store, router, apolloProvider, propsData })
    await Vue.nextTick()

    const selectNone = wrapper.findAll('.select-link').at(0)
    expect(wrapper.find('.el-button--primary').props('disabled')).toBe(true)
    selectNone.trigger('click')
    await Vue.nextTick()

    wrapper.findAll('.el-checkbox').wrappers.forEach((checkBox, index) => {
      expect(checkBox.classes('is-checked')).toBe(false)
    })

    expect(wrapper.find('.el-button--primary').props('disabled')).toBe(false)
  })

  it('it should check all datasets when clicking select all', async() => {
    graphqlWithData()
    const wrapper = mount(testHarness, { store, router, apolloProvider, propsData })
    await Vue.nextTick()

    const selectAll = wrapper.findAll('.select-link').at(1)
    expect(wrapper.find('.el-button--primary').props('disabled')).toBe(true)
    selectAll.trigger('click')
    await Vue.nextTick()

    wrapper.findAll('.el-checkbox').wrappers.forEach((checkBox) => {
      expect(checkBox.classes('is-checked')).toBe(true)
    })

    expect(wrapper.find('.el-button--primary').props('disabled')).toBe(false)
  })

  it('it should check all datasets when clicking select all', async() => {
    graphqlWithData()
    const wrapper = mount(testHarness, { store, router, apolloProvider, propsData })
    await Vue.nextTick()

    const selectAll = wrapper.findAll('.select-link').at(1)
    expect(wrapper.find('.el-button--primary').props('disabled')).toBe(true)
    selectAll.trigger('click')
    await Vue.nextTick()

    wrapper.findAll('.el-checkbox').wrappers.forEach((checkBox) => {
      expect(checkBox.classes('is-checked')).toBe(true)
    })

    expect(wrapper.find('.el-button--primary').props('disabled')).toBe(false)
  })

  it('it should check one dataset and hit update', async() => {
    graphqlWithExtraUserData()

    const wrapper = mount(testHarness, {
      store,
      router,
      apolloProvider,
      propsData,
    })
    await Vue.nextTick()

    expect(wrapper.find('.el-button--primary').props('disabled')).toBe(true)

    expect(wrapper.findAll('.el-checkbox').at(0).classes('is-checked')).toBe(false)

    wrapper.findAll('.el-checkbox').at(0).trigger('click')
    await Vue.nextTick()

    expect(wrapper.findAll('.el-checkbox').at(0).classes('is-checked')).toBe(true)
    expect(wrapper.find('.el-button--primary').props('disabled')).toBe(false)
  })

  it('it should uncheck one dataset and hit update', async() => {
    graphqlWithData()

    const wrapper = mount(testHarness, {
      store,
      router,
      apolloProvider,
      propsData,
    })
    await Vue.nextTick()

    expect(wrapper.find('.el-button--primary').props('disabled')).toBe(true)
    expect(wrapper.findAll('.el-checkbox').at(0).classes('is-checked')).toBe(true)
    wrapper.findAll('.el-checkbox').at(0).trigger('click')
    await Vue.nextTick()

    expect(wrapper.findAll('.el-checkbox').at(0).classes('is-checked')).toBe(false)
    expect(wrapper.find('.el-button--primary').props('disabled')).toBe(false)
  })
})
