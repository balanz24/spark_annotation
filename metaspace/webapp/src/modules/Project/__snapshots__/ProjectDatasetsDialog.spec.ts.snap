// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectDatasetsDialog it should match no dataset snapshot 1`] = `
<transition-stub name="dialog-fade" class="project-datasets-dialog" project="[object Object]" currentuserid="039801c8-919e-11eb-908e-3b2b8e672707" visible="true" ismanager="true" refreshdata="function () {}" style="z-index: 2003;">
  <div class="el-dialog__wrapper">
    <div role="dialog" aria-modal="true" aria-label="Manage Test\`s datasets" class="el-dialog" style="margin-top: 15vh;">
      <div class="el-dialog__header"><span class="el-dialog__title">Manage Test\`s datasets</span><button type="button" aria-label="Close" class="el-dialog__headerbtn"><i class="el-dialog__close el-icon el-icon-close"></i></button></div>
      <div class="el-dialog__body">
        <div class="mt-6">
          <h4 class="m-0">Would you like to include/remove previously submitted datasets?</h4>
          <div class="flex items-center justify-center leading-6">
            <div>No datasets available</div>
          </div>
          <div class="button-bar"><button type="button" class="el-button el-button--default">
              <!---->
              <!----><span>Cancel</span></button><button disabled="disabled" type="button" class="el-button el-button--primary is-disabled w-32">
              <!---->
              <!----><span>Update</span></button></div>
        </div>
      </div>
      <!---->
    </div>
  </div>
</transition-stub>
`;

exports[`ProjectDatasetsDialog it should match snapshot 1`] = `
<transition-stub name="dialog-fade" class="project-datasets-dialog" project="[object Object]" currentuserid="039801c8-919e-11eb-908e-3b2b8e672707" visible="true" ismanager="true" refreshdata="function () {}" style="z-index: 2001;">
  <div class="el-dialog__wrapper">
    <div role="dialog" aria-modal="true" aria-label="Manage Test\`s datasets" class="el-dialog" style="margin-top: 15vh;">
      <div class="el-dialog__header"><span class="el-dialog__title">Manage Test\`s datasets</span><button type="button" aria-label="Close" class="el-dialog__headerbtn"><i class="el-dialog__close el-icon el-icon-close"></i></button></div>
      <div class="el-dialog__body">
        <div class="mt-6">
          <h4 class="m-0">Would you like to include/remove previously submitted datasets?</h4>
          <div class="dataset-checkbox-list leading-6">
            <div class="mb-2"><span class="select-link">Select none</span><span> | </span><span class="select-link">Select all</span></div><label class="el-checkbox is-checked flex h-6 items-center m-0 mx-2"><span class="el-checkbox__input is-checked"><span class="el-checkbox__inner"></span><input type="checkbox" aria-hidden="false" class="el-checkbox__original" value="New 3 (1)"></span><span class="el-checkbox__label"><span class="truncate">New 3 (1)</span><span class="text-gray-700 text-xs tracking-wide pl-1"><mock-elapsed-time date="2021-03-31T14:02:28.722Z"></mock-elapsed-time></span>
              <!----></span></label><label class="el-checkbox is-checked flex h-6 items-center m-0 mx-2"><span class="el-checkbox__input is-checked"><span class="el-checkbox__inner"></span><input type="checkbox" aria-hidden="false" class="el-checkbox__original" value="Untreated_3_434_super_lite_19_31 (1)"></span><span class="el-checkbox__label"><span class="truncate">Untreated_3_434_super_lite_19_31 (1)</span><span class="text-gray-700 text-xs tracking-wide pl-1"><mock-elapsed-time date="2021-03-30T21:25:18.473Z"></mock-elapsed-time></span>
              <!----></span></label>
          </div>
          <div class="button-bar"><button type="button" class="el-button el-button--default">
              <!---->
              <!----><span>Cancel</span></button><button disabled="disabled" type="button" class="el-button el-button--primary is-disabled w-32">
              <!---->
              <!----><span>Update</span></button></div>
        </div>
      </div>
      <!---->
    </div>
  </div>
</transition-stub>
`;
